src/adapter/src/rbac.rs

```diff
@@ -0,0 +1,236 @@
+use crate::session::Session;
+use crate::AdapterError;
+use mz_sql::catalog::{CatalogRole, SessionCatalog};
+use mz_sql::plan::Plan;
+use std::error::Error;
+use std::fmt;
+use std::fmt::Formatter;
+
+/// Attributes that allow a role to execute certain plans.
+///
+/// Note: This is a subset of all role attributes used for privilege checks.
+#[derive(Debug)]
+pub enum Attribute {
+    CreateRole,
+    CreateDB,
+    CreateCluster,
+}
+
+impl Attribute {
+    /// Checks if a role has the privilege granted by the attribute.
+    fn check_role(&self, role: &dyn CatalogRole) -> bool {
+        match self {
+            Attribute::CreateRole => role.create_role(),
+            Attribute::CreateDB => role.create_db(),
+            Attribute::CreateCluster => role.create_cluster(),
+        }
+    }
+}
+
+impl ToString for Attribute {
+    fn to_string(&self) -> String {
+        match self {
+            Attribute::CreateRole => "CREATEROLE".into(),
+            Attribute::CreateDB => "CREATEDB".into(),
+            Attribute::CreateCluster => "CREATECLUSTER".into(),
+        }
+    }
+}
+
+/// Errors that can occur due to an unauthorized action.
+#[derive(Debug)]
+pub struct UnauthorizedError {
+    action: String,
+    kind: UnauthorizedErrorKind,
+}
+
+#[derive(Debug)]
+enum UnauthorizedErrorKind {
+    Superuser,
+    Attribute(Attribute),
+    // TODO(jkosh44) replace with structured errors once RBAC is done.
+    Unstructured { reason: String },
+}
+
+impl fmt::Display for UnauthorizedError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "permission denied to {}", self.action)
+    }
+}
+
+impl UnauthorizedError {
+    pub fn superuser(action: String) -> UnauthorizedError {
+        UnauthorizedError {
+            action,
+            kind: UnauthorizedErrorKind::Superuser,
+        }
+    }
+
+    pub fn attribute(action: String, attribute: Attribute) -> UnauthorizedError {
+        UnauthorizedError {
+            action,
+            kind: UnauthorizedErrorKind::Attribute(attribute),
+        }
+    }
+
+    pub fn unstructured(action: String, reason: String) -> UnauthorizedError {
+        UnauthorizedError {
+            action,
+            kind: UnauthorizedErrorKind::Unstructured { reason },
+        }
+    }
+
+    pub fn detail(&self) -> String {
+        match &self.kind {
+            UnauthorizedErrorKind::Superuser => {
+                format!("You must be a superuser to {}", self.action)
+            }
+            UnauthorizedErrorKind::Attribute(attribute) => format!(
+                "You must have the {} attribute to {}",
+                attribute.to_string(),
+                self.action
+            ),
+            UnauthorizedErrorKind::Unstructured { reason } => {
+                format!("{} to {}", reason, self.action)
+            }
+        }
+    }
+}
+
+impl Error for UnauthorizedError {}
+
+pub fn check_command(catalog: &Catalog, cmd: &Command) -> Result<(), UnauthorizedError> {
```

Comment:
All the check functions need doc comments, especially because it's tricky to know how to review this for correctness without knowing its intention.

Reply:
Done